r/content/datasets.Rmd

```diff
@@ -0,0 +1,291 @@
+  partitioning = "homeworld")
+```
+
+
+```{r, test_write_dataset_partitioned, opts.label = "test"}
+
+test_that("write_dataset_partitioned works as expected", {
+  
+  expect_true(file.exists("starwars_data_partitioned"))
+  expect_length(list.files("starwars_data_partitioned", recursive = TRUE), 49)
+  
+})
+
+```
+
+### Discussion
+
+The data is written to separate folders based on the values in the `homeworld` 
+column.  The default behaviour is to use Hive-style (i.e. "col_name=value" folder names)
+partitions.
+
+```{r}
+# Take a look at the files in this directory
+list.files("starwars_data_partitioned", recursive = TRUE)
+```
+
+Note that in the example above, when there was an `NA` value in the `homeworld`
+column, these values are written to the `homeworld=__HIVE_DEFAULT_PARTITION__`
+directory.
+
+You can specify multiple partitioning variables to add extra levels of partitioning.
+
+```{r, write_dataset_partitioned_deeper}
+write_dataset(
+  dataset = starwars,
+  path = "starwars_partitioned_twice",
+  partitioning = c("homeworld", "species")
+)
+```
+
+```{r, test_write_dataset_partitioned_deeper, opts.label = "test"}
+
+test_that("write_dataset_partitioned_deeper works as expected", {
+  
+  expect_true(file.exists("starwars_partitioned_twice"))
+  expect_length(list.files("starwars_partitioned_twice", recursive = TRUE), 58)
+  
+})
+
+```
+
+```{r}
+# Take a look at the files in this directory
+list.files("starwars_partitioned_twice", recursive = TRUE)
+```
+
+There are two different ways to specify variables to use for partitioning - 
+either via the `partitioning` variable as above, or by using `dplyr::group_by()` on your data - the group variables will form the partitions.
+
+```{r, write_dataset_partitioned_groupby}
+library(dplyr)
+write_dataset(dataset = group_by(starwars, homeworld, species),
+  path = "starwars_groupby")
+```
+
+```{r, test_write_dataset_partitioned_groupby, opts.label = "test"}
+
+test_that("write_dataset_partitioned_groupby works as expected", {
+  
+  expect_true(file.exists("starwars_groupby"))
+  expect_length(list.files("starwars_groupby", recursive = TRUE), 58)
+  
+})
+
+```
+
+```{r}
+# Take a look at the files in this directory
+list.files("starwars_groupby", recursive = TRUE)
+```
+## Write data to disk - Feather format
+
+You want to write data to disk in a single Feather file.
+
+### Solution
+
+```{r, write_dataset_feather}
+write_dataset(dataset = starwars,
+  path = "starwars_data_feather",
+  format = "feather")
+```
+
+```{r, test_write_dataset_feather, opts.label = "test"}
+
+test_that("write_dataset_feather works as expected", {
+  
+  expect_true(file.exists("starwars_data_feather"))
+  expect_length(list.files("starwars_data_feather"), 1)
+  
+})
+
+```
+
+## Write data to disk - CSV format
+
+You want to write data to disk in a single CSV file.
+
+### Solution
+
+```{r, write_dataset_csv}
+# Need to update this example as we can't write list columns to CSV :(
```

Comment:
How about we either make a subset of starwars dataset at the start of this chapter/section to use later, which doesn't include the list column, or just acknowledge the list column issue in the discussion section?

Reply:
I refactored to stick with the `airquality` dataset, to be more consistent with the rest fo the read/write material in the R cookbook.